freetextresponse/freetextresponse.py

```diff
@@ -18,11 +18,69 @@
 from xblock.fragment import Fragment
 from xblock.validation import ValidationMessage
 from xblockutils.studio_editable import StudioEditableXBlockMixin
+from xblockutils.publish_event import PublishEventMixin
 from .mixins import EnforceDueDates
 
+from answer_pool import offer_answer, get_other_answers
+
+class MissingDataFetcherMixin:
+    """
+    The mixin used for generating student item dict for submission API
+
+    Copied from
+    https://github.com/edx/edx-ora2/blob/master/openassessment/xblock/openassessmentblock.py
+    """
+    def get_course_id(self):
+        return self._serialize_opaque_key(self.xmodule_runtime.course_id)
+
+    def get_student_item_dict(self, anonymous_user_id=None):
+        """
+        Create a student_item_dict from our surrounding context.
+
+        See also: submissions.api for details.
+
+        Args:
+            anonymous_user_id(str): A unique anonymous_user_id for (user, course) pair.
+        Returns:
+            (dict): The student item associated with this XBlock instance. This includes the
+                student id, item id, and course id.
+        """
+        item_id = self._serialize_opaque_key(self.scope_ids.usage_id)
+
+        if hasattr(self, "xmodule_runtime"):
+            course_id = self.get_course_id()  # pylint:disable=E1101
+            if anonymous_user_id:
+                student_id = anonymous_user_id
+            else:
+                student_id = self.xmodule_runtime.anonymous_student_id  #pylint:disable=E1101
+        else:
+            course_id = "edX/Enchantment_101/April_1"
+            if self.scope_ids.user_id is None:
```

Comment:
I think you could simplify this like:

```py
student_id = self.scope_ids.user_id or ''
student_id = unicode(student_id)
```

since both `0` and the empty string `''` are invalid student ids.